File: script/menu.js
const isMac = process.platform === 'darwin'

let buildFileMenu = () => {
	let fileMenu = new nw.Menu()

	exports.newSprite = new nw.MenuItem({
		label: 'New',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'n',
		click: () => nw.Window.get().window.sketch.newSprite()
	})
	fileMenu.append(exports.newSprite)

	exports.openSprite = new nw.MenuItem({
		label: 'Open...',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'o',
		click: () => nw.Window.get().window.sketch.openSprite()
	})
	fileMenu.append(exports.openSprite)

	fileMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.saveSprite = new nw.MenuItem({
		label: 'Save',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 's',
		enabled: false,
		click: () => nw.Window.get().window.sketch.saveSprite()
	})
	fileMenu.append(exports.saveSprite)

	exports.saveAsSprite = new nw.MenuItem({
		label: 'Save As...',
		modifiers: isMac ? 'shift+cmd' : 'ctrl+shift',
		key: 's',
		enabled: false,
		click: () => nw.Window.get().window.sketch.saveAsSprite()
	})
	fileMenu.append(exports.saveAsSprite)

	fileMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.exportSPR = new nw.MenuItem({
		label: 'Export As SPR...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportSPR()
	})
	fileMenu.append(exports.exportSPR)

	exports.exportS16 = new nw.MenuItem({
		label: 'Export As S16...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportS16()
	})
	fileMenu.append(exports.exportS16)

	exports.exportC16 = new nw.MenuItem({
		label: 'Export As C16...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportC16()
	})
	fileMenu.append(exports.exportC16)

	exports.exportBLK = new nw.MenuItem({
		label: 'Export As BLK...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportBLK()
	})
	fileMenu.append(exports.exportBLK)

	fileMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.exportPNG = new nw.MenuItem({
		label: 'Export As PNG...',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'e',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportPNG()
	})
	fileMenu.append(exports.exportPNG)

	exports.exportGIF = new nw.MenuItem({
		label: 'Export As GIF...',
		modifiers: isMac ? 'shift+cmd' : 'ctrl+shift',
		key: 'e',
		enabled: false,
		click: () => nw.Window.get().window.sketch.exportGIF()
	})
	fileMenu.append(exports.exportGIF)

	return fileMenu
}

let buildEditMenu = () => {
	let editMenu = new nw.Menu()

	exports.undo = new nw.MenuItem({
		label: 'Undo',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'z',
		enabled: false,
		click: () => nw.Window.get().window.sketch.undo()
	})
	editMenu.append(exports.undo)

	exports.redo = new nw.MenuItem({
		label: 'Redo',
		modifiers: isMac ? 'shift+cmd' : 'ctrl+shift',
		key: 'z',
		enabled: false,
		click: () => nw.Window.get().window.sketch.redo()
	})
	editMenu.append(exports.redo)

	editMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.cutFrame = new nw.MenuItem({
		label: 'Cut',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'x',
		enabled: false,
		click: () => nw.Window.get().window.sketch.cutFrame()
	})
	editMenu.append(exports.cutFrame)

	exports.copyFrame = new nw.MenuItem({
		label: 'Copy',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'c',
		enabled: false,
		click: () => nw.Window.get().window.sketch.copyFrame()
	})
	editMenu.append(exports.copyFrame)

	exports.pasteFrame = new nw.MenuItem({
		label: 'Paste',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'v',
		enabled: false,
		click: () => nw.Window.get().window.sketch.pasteFrame()
	})
	editMenu.append(exports.pasteFrame)

	exports.deleteFrame = new nw.MenuItem({
		label: 'Delete',
		key: 'Delete',
		enabled: false,
		click: () => nw.Window.get().window.sketch.deleteFrame()
	})
	editMenu.append(exports.deleteFrame)

	editMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.selectAllFrames = new nw.MenuItem({
		label: 'Select All',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'a',
		enabled: false,
		click: () => nw.Window.get().window.sketch.selectAllFrames()
	})
	editMenu.append(exports.selectAllFrames)

	exports.deselectAllFrames = new nw.MenuItem({
		label: 'Deselect',
		modifiers: isMac ? 'shift+cmd' : 'ctrl+shift',
		key: 'a',
		enabled: false,
		click: () => nw.Window.get().window.sketch.deselectAllFrames()
	})
	editMenu.append(exports.deselectAllFrames)

	editMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.insertImage = new nw.MenuItem({
		label: 'Insert Image...',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: 'i',
		enabled: false,
		click: () => nw.Window.get().window.sketch.insertImage()
	})
	editMenu.append(exports.insertImage)

	return editMenu
}

let buildViewMenu = () => {
	let viewMenu = new nw.Menu()

	exports.resetZoom = new nw.MenuItem({
		label: '100%',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: '0',
		enabled: false,
		click: () => nw.Window.get().window.sketch.resetZoom()
	})
	viewMenu.append(exports.resetZoom)

	exports.zoomIn = new nw.MenuItem({
		label: 'Zoom In',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: '=',
		enabled: false,
		click: () => nw.Window.get().window.sketch.zoomIn()
	})
	viewMenu.append(exports.zoomIn)

	exports.zoomOut = new nw.MenuItem({
		label: 'Zoom Out',
		modifiers: isMac ? 'cmd' : 'ctrl',
		key: '-',
		enabled: false,
		click: () => nw.Window.get().window.sketch.zoomOut()
	})
	viewMenu.append(exports.zoomOut)

	viewMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.loadPalette = new nw.MenuItem({
		label: 'Load Palette...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.loadPalette()
	})
	viewMenu.append(exports.loadPalette)

	exports.convertPalette = new nw.MenuItem({
		label: 'Convert to Palette...',
		enabled: false,
		click: () => nw.Window.get().window.sketch.convertPalette()
	})
	viewMenu.append(exports.convertPalette)

	exports.resetPalette = new nw.MenuItem({
		label: 'Reset Palette',
		enabled: false,
		click: () => nw.Window.get().window.sketch.resetPalette()
	})
	viewMenu.append(exports.resetPalette)

	viewMenu.append(new nw.MenuItem({ type: 'separator' }))

	exports.viewAsSprite = new nw.MenuItem({
		label: 'View As Sprite',
		type: 'checkbox',
		checked: true,
		enabled: false,
		click: () => nw.Window.get().window.sketch.viewAsSprite()
	})
	viewMenu.append(exports.viewAsSprite)

	exports.viewAsBackground = new nw.MenuItem({
		label: 'View As Background',
		type: 'checkbox',
		enabled: false,
		click: () => nw.Window.get().window.sketch.viewAsBackground()
	})
	viewMenu.append(exports.viewAsBackground)

	viewMenu.append(new nw.MenuItem({ type: 'separator' }))

	let transparentMenu = new nw.Menu()
	exports.setTransparentBlack = new nw.MenuItem({
		label: 'Black',
		type: 'checkbox',
		checked: true,
		click: () => nw.Window.get().window.api.setTransparentColor('black')
	})
	transparentMenu.append(exports.setTransparentBlack)
	exports.setTransparentWhite = new nw.MenuItem({
		label: 'White',
		type: 'checkbox',
		click: () => nw.Window.get().window.api.setTransparentColor('white')
	})
	transparentMenu.append(exports.setTransparentWhite)
	exports.setTransparentNone = new nw.MenuItem({
		label: 'Transparent',
		type: 'checkbox',
		click: () => nw.Window.get().window.api.setTransparentColor('none')
	})
	transparentMenu.append(exports.setTransparentNone)
	exports.setTransparentColor = new nw.MenuItem({
		label: 'Transparent Color',
		submenu: transparentMenu
	})
	viewMenu.append(exports.setTransparentColor)

	return viewMenu
}

exports.buildMenu = () => {
	let menu = new nw.Menu({ type: 'menubar' })

	let fileMenu = new nw.MenuItem({
		label: 'File',
		submenu: buildFileMenu()
	})

	let editMenu = new nw.MenuItem({
		label: 'Edit',
		submenu: buildEditMenu()
	})

	let viewMenu = new nw.MenuItem({
		label: 'View',
		submenu: buildViewMenu()
	})

	if (isMac) {
		menu.createMacBuiltin('Spritist', { hideEdit: true, hideWindow: false })
		menu.insert(fileMenu, 1)
		menu.insert(editMenu, 2)
		menu.insert(viewMenu, 3)
	} else {
		menu.append(fileMenu)
		menu.append(editMenu)
		menu.append(viewMenu)
	}

	return menu
}

File: src-tauri/src/menu.rs
use tauri::{
	CustomMenuItem,
	Menu,
	MenuItem,
	Submenu
};

pub fn build_menu() -> Menu {
	let new_file = CustomMenuItem::new("new_file".to_string(), "New").accelerator("CmdOrCtrl+N");
	let open_file = CustomMenuItem::new("open_file".to_string(), "Open...").accelerator("CmdOrCtrl+O");
	// let open_recent = Submenu::new("Open Recent", Menu::new());
	let save_file = CustomMenuItem::new("save_file".to_string(), "Save").accelerator("CmdOrCtrl+S").disabled();
	let save_as = CustomMenuItem::new("save_as".to_string(), "Save As...").accelerator("CmdOrCtrl+Shift+S").disabled();
	let import_menu = Submenu::new("Import", Menu::new()
		.add_item(CustomMenuItem::new("import_png_as_blk".to_string(), "Import PNG as BLK").accelerator("CmdOrCtrl+Shift+B"))
		.add_item(CustomMenuItem::new("import_spritesheet".to_string(), "Import Spritesheet").accelerator("CmdOrCtrl+Shift+T"))
	);
	let export_menu = Submenu::new("Export", Menu::new()
		.add_item(CustomMenuItem::new("export_png".to_string(), "Export PNG").accelerator("CmdOrCtrl+E").disabled())
		.add_item(CustomMenuItem::new("export_gif".to_string(), "Export GIF").accelerator("CmdOrCtrl+G").disabled())
		.add_item(CustomMenuItem::new("export_spritesheet".to_string(), "Export Spritesheet").accelerator("CmdOrCtrl+T").disabled())
	);
	let file_menu = Submenu::new("File", Menu::new()
		.add_item(new_file)
		.add_item(open_file)
		// .add_submenu(open_recent)
		.add_native_item(MenuItem::Separator)
		.add_item(save_file)
		.add_item(save_as)
		.add_native_item(MenuItem::Separator)
		.add_submenu(import_menu)
		.add_submenu(export_menu)
		.add_native_item(MenuItem::Separator)
		.add_native_item(MenuItem::Quit)
	);

	let undo = CustomMenuItem::new("undo".to_string(), "Undo").accelerator("CmdOrCtrl+Z").disabled();
	let redo = CustomMenuItem::new("redo".to_string(), "Redo").accelerator("CmdOrCtrl+Y").disabled();
	let cut = CustomMenuItem::new("cut".to_string(), "Cut").accelerator("CmdOrCtrl+X").disabled();
	let copy = CustomMenuItem::new("copy".to_string(), "Copy").accelerator("CmdOrCtrl+C").disabled();
	let paste = CustomMenuItem::new("paste".to_string(), "Paste").accelerator("CmdOrCtrl+V").disabled();
	let delete = CustomMenuItem::new("delete".to_string(), "Delete").accelerator("Delete").disabled();
	let select_all = CustomMenuItem::new("select_all".to_string(), "Select All").accelerator("CmdOrCtrl+A");
	let deselect_all = CustomMenuItem::new("deselect_all".to_string(), "Deselect All").accelerator("CmdOrCtrl+D").disabled();
	let insert_image = CustomMenuItem::new("insert_image".to_string(), "Insert Image...").accelerator("CmdOrCtrl+I").disabled();
	let replace_frame = CustomMenuItem::new("replace_frame".to_string(), "Replace Frame...").accelerator("CmdOrCtrl+Shift+I").disabled();
	let edit_menu = Submenu::new("Edit", Menu::new()
		.add_item(undo)
		.add_item(redo)
		.add_native_item(MenuItem::Separator)
		.add_item(cut)
		.add_item(copy)
		.add_item(paste)
		.add_item(delete)
		.add_native_item(MenuItem::Separator)
		.add_item(select_all)
		.add_item(deselect_all)
		.add_native_item(MenuItem::Separator)
		.add_item(insert_image)
		.add_item(replace_frame)
	);

	let reset_zoom = CustomMenuItem::new("reset_zoom".to_string(), "100%").accelerator("CmdOrCtrl+0").disabled();
	let zoom_in = CustomMenuItem::new("zoom_in".to_string(), "Zoom In").accelerator("CmdOrCtrl+=");
	let zoom_out = CustomMenuItem::new("zoom_out".to_string(), "Zoom Out").accelerator("CmdOrCtrl+-").disabled();
	let palette_menu = Submenu::new("SPR Palette", Menu::new()
		.add_item(CustomMenuItem::new("load_original".to_string(), "✔ Original Palette"))
		.add_item(CustomMenuItem::new("load_reversed".to_string(), "- Reversed Palette"))
		.add_item(CustomMenuItem::new("load_palette".to_string(), "- Custom Palette..."))
		.add_native_item(MenuItem::Separator)
		.add_item(CustomMenuItem::new("convert_to_original".to_string(), "Convert to Original Palette").disabled())
		.add_item(CustomMenuItem::new("convert_to_reversed".to_string(), "Convert to Reversed Palette").disabled())
		.add_item(CustomMenuItem::new("convert_to_palette".to_string(), "Convert to Palette...").disabled())
	);
	let view_as_sprite = CustomMenuItem::new("view_as_sprite".to_string(), "✔ View As Sprite").disabled();
	let view_as_bg = CustomMenuItem::new("view_as_bg".to_string(), "- View As Background").disabled();
	let show_image_info = CustomMenuItem::new("show_image_info".to_string(), "✔ Show Image Info");
	let transparent_menu = Submenu::new("Transparent Color", Menu::new()
		.add_item(CustomMenuItem::new("transparent_black".to_string(), "✔ Black"))
		.add_item(CustomMenuItem::new("transparent_white".to_string(), "- White"))
		.add_item(CustomMenuItem::new("transparent_none".to_string(), "- Transparent"))
	);
	let theme_menu = Submenu::new("Theme", Menu::new()
		.add_item(CustomMenuItem::new("theme_dark".to_string(), "✔ Dark"))
		.add_item(CustomMenuItem::new("theme_light".to_string(), "- Light"))
		.add_item(CustomMenuItem::new("theme_purple".to_string(), "- Purple"))
	);
	let view_menu = Submenu::new("View", Menu::new()
		.add_item(reset_zoom)
		.add_item(zoom_in)
		.add_item(zoom_out)
		.add_native_item(MenuItem::Separator)
		.add_submenu(palette_menu)
		.add_native_item(MenuItem::Separator)
		.add_item(view_as_sprite)
		.add_item(view_as_bg)
		.add_native_item(MenuItem::Separator)
		.add_item(show_image_info)
		.add_submenu(transparent_menu)
		.add_submenu(theme_menu)
	);

	Menu::new()
		.add_submenu(file_menu)
		.add_submenu(edit_menu)
		.add_submenu(view_menu)
}
